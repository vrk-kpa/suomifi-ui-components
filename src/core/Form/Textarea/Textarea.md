The `<Textarea>` component is suitable for longer user inputs.

### Accessibility Notes

- Success/error states are not distinguishable without colour.

Examples:

- [Basic use](./#/Components/Textarea?id=basic-use)
- [Hint text](./#/Components/Textarea?id=hint-text)
- [Default value](./#/Components/Textarea?id=default-value)
- [Controlled state](./#/Components/Textarea?id=controlled-state)
- [Error status](./#/Components/Textarea?id=error-status)
- [Disabled](./#/Components/Textarea?id=disabled)
- [Optional input](./#/Components/Textarea?id=optional-input)
- [Full width](./#/Components/Textarea?id=full-width)
- [Resize modes](./#/Components/Textarea?id=resize-modes)
- [Tooltip](./#/Components/Textarea?id=tooltip)
- [Character counter](./#/Components/Textarea?id=character-counter)

<div style="margin-bottom: 40px">
  [Props & methods](./#/Components/Textarea?id=props--methods)
</div>

### Basic use

Provide a descriptive `labelText` for the input

```js
import { Textarea } from 'suomifi-ui-components';

<Textarea labelText="Additional information" />;
```

### Hint text

A `hintText` can be given to the component to provide a better explanation of the purpose and requirements of the input.

```js
import { Textarea } from 'suomifi-ui-components';

<Textarea
  labelText="Additional information"
  hintText="Provide any other details you want to add to your application"
/>;
```

### Default value

Set the component's initial value by providing text as children

```js
import { TextInput } from 'suomifi-ui-components';

<Textarea
  labelText="Additional information"
  hintText="Provide any other details you want to add to your application"
>
  Lorem ipsum dolor sit amet
</Textarea>;
```

### Controlled state

Access and control the input's value programmatically with the `value` prop.

A typical use case involves setting the state in the `onChange()` function.

```js
import { Textarea } from 'suomifi-ui-components';
import { useState } from 'react';

const [inputValue, setInputValue] = useState('');

<Textarea
  onChange={(event) => setInputValue(event.target.value)}
  labelText="Additional information"
  value={inputValue}
/>;
```

### Error status

Control the error status of the component using the `status` and `statusText` props.

```js
import { Textarea } from 'suomifi-ui-components';

const [errorState, setErrorState] = React.useState(true);
const statusText = errorState
  ? 'Description is a required field'
  : undefined;
const status = errorState ? 'error' : 'default';

<Textarea
  labelText="Description"
  statusText={statusText}
  status={status}
  onChange={(event) => setErrorState(event.target.value === '')}
/>;
```

### Disabled

Disable the input with the `disabled` prop.

```js
import { Textarea } from 'suomifi-ui-components';

<Textarea labelText="Additional information" disabled />;
```

### Optional input

Suomi.fi inputs are required by default, but can be marked optional using the `optionalText` property.

```js
import { Textarea } from 'suomifi-ui-components';

<Textarea
  labelText="Additional information"
  optionalText="optional"
/>;
```

### Full width

The component can be made to take all available horizontal space by giving it the `fullWidth` property.

```js
import { Textarea } from 'suomifi-ui-components';

<Textarea
  labelText="Additional information"
  hintText="Provide any other details you want to add to your application"
  fullWidth
/>;
```

### Resize modes

Textarea can be resized vertically by default. A `resize` property can be given to control the ways the component can be resized by the user.

To allow horizontal or free resize, you must override the default container width either by styling or using the `fullWidth` prop.

```js
import { Textarea } from 'suomifi-ui-components';
<Textarea
  labelText="Additional information"
  resize="both"
  fullWidth
/>;
```

### Tooltip

A `<Tooltip>` component can be used with Textarea to provide additional information.

In terms of instructive texts, Tooltip should only be used as a "last resort" when the info text is too long for `hintText`. Tooltip can be used for other nice-to-know information.

For instructions regarding how to ensure your Tooltip is accessible, please refer to the [Tooltip documentation](./#/Components/Tooltip).

```js
import {
  Textarea,
  Tooltip,
  Heading,
  Text
} from 'suomifi-ui-components';

const labelTextForTooltipExample = 'Additional feedback';

<Textarea
  labelText={labelTextForTooltipExample}
  fullWidth
  tooltipComponent={
    <Tooltip
      ariaToggleButtonLabelText={`${labelTextForTooltipExample}, additional information`}
      ariaCloseButtonLabelText={`${labelTextForTooltipExample}, close additional information`}
    >
      <Heading variant="h5" as="h2">
        Additional feedback
      </Heading>
      <Text>Consider answering some of the following questions:</Text>
      <ul>
        <li>How was your overall experience</li>
        <li>What went well</li>
        <li>What things could've gone better and how?</li>
      </ul>
      <Text>
        We use this information in order to improve our service and it
        will not be linked to your profile.
      </Text>
    </Tooltip>
  }
/>;
```

### Character counter

- Provide the maximum length of the text to enable character counter
- Provide clear texts indicating characters remaining or exceeding the limit. They are hidden from view but exist for screen reader users.
- There is an inbuilt delay in updating the status text to make it work better for VoiceOver users

```js
import {
  Textarea,
  Tooltip,
  Heading,
  Text
} from 'suomifi-ui-components';
import React, { useState } from 'react';

const [isError, setIsError] = useState(false);
const [statusText, setStatusText] = useState('');

const maxCharAmount = 30;

/**
 * Check if maximum amount of characters has exceed, and set status and statusText accordingly.
 * You can also add any other desired input validation rules here.
 */
const validateText = (text) => {
  if (text.length > maxCharAmount) {
    setIsError(true);
    setStatusText('Description must be 30 characters or less');
  } else {
    setIsError(false);
    setStatusText('');
  }
};

<>
  <Textarea
    hintText="Please provide details pertaining to the case"
    labelText="Additional details"
    onChange={() => validateText(event.target.value)}
    characterLimit={maxCharAmount}
    ariaCharactersRemainingText={(amount) =>
      `You have ${amount} characters remaining`
    }
    ariaCharactersExceededText={(amount) =>
      `You have ${amount} characters too many`
    }
    statusText={statusText}
    status={isError ? 'error' : 'default'}
    fullWidth
  >
    Lorem ipsum dolor sit amet
  </Textarea>
</>;
```

### Props & methods

Textarea component supports [margin props](./#/Spacing/Margin%20props) for spacing.
