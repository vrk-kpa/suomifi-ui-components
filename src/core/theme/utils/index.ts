export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
