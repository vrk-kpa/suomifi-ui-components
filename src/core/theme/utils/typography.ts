import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
