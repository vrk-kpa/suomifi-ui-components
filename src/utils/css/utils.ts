import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
