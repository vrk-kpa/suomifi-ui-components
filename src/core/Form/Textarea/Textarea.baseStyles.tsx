import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { element, fixInternalMargins, font } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${font(theme)('bodyText')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)};
  ${fixInternalMargins()}
  color: ${theme.colors.blackBase};
  width: 290px;

  &.fi-textarea {
    display: flex;
    flex-direction: column;

    & .fi-textarea_bottom-wrapper {
      display: flex;
      justify-content: space-between;
    }

    & .fi-textarea_character-counter {
      ${font(theme)('bodyTextSmall')};
      color: ${theme.colors.blackBase};
      font-size: 14px;
      line-height: 20px;
      flex: none;
      margin-top: ${theme.spacing.xxs};

      &.fi-textarea_character-counter--error {
        color: ${theme.colors.alertBase};
        ${font(theme)('bodySemiBoldSmall')};
        font-size: 14px;
        line-height: 20px;
      }
    }

    & .fi-textarea_statusText--has-content {
      margin-top: ${theme.spacing.xxs};
    }

    & .fi-textarea_textarea-element-container {
      &:focus-within {
        ${theme.focuses.highContrastFocus} /* For high contrast mode */
        position: relative;

        &::after {
          ${theme.focuses.absoluteFocus}
        }
      }
    }

    & .fi-label {
      margin-bottom: ${theme.spacing.xs};
    }

    & .fi-hint-text {
      margin-bottom: ${theme.spacing.xs};
    }

    & .fi-textarea_textarea {
      display: block;
      resize: vertical;
      border-radius: 2px;
      border: 1px solid ${theme.colors.depthDark3};
      box-shadow: ${theme.shadows.actionElementBoxShadow};
      padding: 8px 14px 13px 10px;
      ${theme.typography.bodyTextSmall};
      width: 100%;
      background-color: ${theme.colors.whiteBase};

      &:focus {
        outline: none;
      }

      &::placeholder {
        font-style: italic;
        color: ${theme.colors.depthDark2};
        opacity: 1;
      }

      &-resize--horizontal {
        resize: horizontal;
      }

      &-resize--both {
        resize: both;
      }

      &-resize--none {
        resize: none;
      }
    }

    & .fi-status-text {
      display: block;
      line-height: 18px;
    }

    &.fi-textarea--disabled {
      color: ${theme.colors.depthBase};
      cursor: not-allowed;

      & .fi-textarea_textarea {
        background-color: ${theme.colors.depthLight3};
      }
    }

    &.fi-textarea--error {
      & .fi-textarea_textarea {
        border: 2px solid ${theme.colors.alertBase};
      }
    }
  }

  &.fi-textarea--full-width {
    width: 100%;
  }
`;
