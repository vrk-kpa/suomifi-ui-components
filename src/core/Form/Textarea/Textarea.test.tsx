import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { Textarea } from './Textarea';

describe('snapshot', () => {
  test('default structure should match snapshot', () => {
    const { container } = render(
      <Textarea id="just-for-snapshot-to-be-same" labelText="Label text" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('props', () => {
  describe('default structure, with default props', () => {
    const DefaultTextareaComponent = <Textarea labelText="Label here" />;

    it('should have baseClassName', () => {
      const { container } = render(DefaultTextareaComponent);
      expect(container.firstChild).toHaveClass('fi-textarea');
    });

    it('should have <textarea< HTML-element with correct class ', () => {
      const { getByRole } = render(DefaultTextareaComponent);
      const textarea = getByRole('textbox');
      expect(textarea).toHaveClass('fi-textarea_textarea');
    });

    it('should have label text with correct class', () => {
      const { getByText } = render(DefaultTextareaComponent);
      const label = getByText('Label here');
      expect(label).toHaveClass('fi-label_label-span');
    });

    it('has user given aria-describedby on textarea', () => {
      const { getByRole } = render(
        <Textarea
          labelText="Label here"
          aria-describedby="external-component-id"
        />,
      );
      expect(getByRole('textbox')).toHaveAttribute(
        'aria-describedby',
        'external-component-id',
      );
    });

    it(
      'should not have basic accessibility issues',
      axeTest(DefaultTextareaComponent),
    );
  });

  describe('className', () => {
    it('should have the given custom classname ', () => {
      const { container } = render(
        <Textarea labelText="label" className="custom-classname" />,
      );
      expect(container.firstChild).toHaveClass('custom-classname');
    });
  });

  describe('disabled', () => {
    it('should have disabled attribute and correct class', () => {
      const { container, getByRole } = render(
        <Textarea labelText="label" disabled />,
      );
      expect(container.firstChild).toHaveClass('fi-textarea--disabled');
      const textarea = getByRole('textbox');
      expect(textarea).toHaveAttribute('disabled');
    });
  });

  describe('onBlur', () => {
    test('should notice when leaving area', () => {
      const mockOnBlur = jest.fn();
      const { getByRole } = render(
        <Textarea labelText="label" onBlur={mockOnBlur} />,
      );
      const textarea = getByRole('textbox');
      fireEvent.blur(textarea);
      expect(mockOnBlur).toBeCalledTimes(1);
    });
  });

  describe('onClick', () => {
    it('should notice click', () => {
      const mockOnClick = jest.fn();
      const { getByRole } = render(
        <Textarea labelText="label" onClick={mockOnClick} />,
      );
      const textarea = getByRole('textbox');
      fireEvent.mouseDown(textarea);
      expect(mockOnClick).toBeCalledTimes(1);
    });
  });

  describe('onChange', () => {
    it('should notice change and have the given text', () => {
      const mockOnChange = jest.fn();
      const { getByRole } = render(
        <Textarea labelText="label" onChange={mockOnChange} />,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      fireEvent.change(textarea, { target: { value: 'abc' } });
      expect(mockOnChange).toBeCalledTimes(1);
      expect(textarea.value).toBe('abc');
    });
  });

  describe('labelText', () => {
    it('should have the given text on start and after changing to new', () => {
      const { rerender, queryByText } = render(
        <Textarea labelText="Summer time" />,
      );
      expect(queryByText('Summer time')).not.toBeNull();
      rerender(<Textarea labelText="Winter is coming" />);
      expect(queryByText('Winter is coming')).not.toBeNull();
    });
  });

  describe('labelMode', () => {
    it('hidden: should hide label; label to have visually-hidden classname', () => {
      const { getByText } = render(
        <Textarea labelText="To be hidden" labelMode="hidden" />,
      );
      const label = getByText('To be hidden');
      expect(label).toHaveClass('fi-visually-hidden');
    });
  });

  describe('visualPlaceholder', () => {
    it('should have the given text as attribute', () => {
      const { getByRole } = render(
        <Textarea
          labelText="To be hidden"
          visualPlaceholder="Enter text here"
        />,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('placeholder', 'Enter text here');
    });
  });

  describe('children', () => {
    it('should have given children as content for textarea', () => {
      const { getByRole } = render(
        <Textarea labelText="label">Text written!</Textarea>,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      expect(textarea).toHaveTextContent('Text written!');
    });
  });

  describe('hintText', () => {
    it('should have element and correct classname for it', () => {
      const { getByText } = render(
        <Textarea labelText="label" hintText="Example hint text" />,
      );
      const hintText = getByText('Example hint text');
      expect(hintText).toHaveClass('fi-hint-text');
    });

    it('will be added to input aria-describedby', () => {
      const { getByRole } = render(
        <Textarea id="123" labelText="label" hintText="Example hint text" />,
      );
      expect(getByRole('textbox')).toHaveAttribute(
        'aria-describedby',
        '123-hintText',
      );
    });
  });

  describe('statusText', () => {
    it('should have element and correct classname for it', () => {
      const { getByText } = render(
        <Textarea labelText="label" statusText="EROR EROR" />,
      );
      const statusText = getByText('EROR EROR');
      expect(statusText).toHaveClass('fi-status-text');
    });

    it('will be added to input aria-describedby', () => {
      const { getByRole } = render(
        <Textarea
          id="123"
          labelText="label"
          statusText="Example status text"
        />,
      );
      expect(getByRole('textbox')).toHaveAttribute(
        'aria-describedby',
        '123-statusText',
      );
    });
  });

  describe('status', () => {
    it('should have error classname', () => {
      const { container } = render(
        <Textarea labelText="label" status="error" statusText="EROR EROR" />,
      );
      expect(container.firstChild).toHaveClass('fi-textarea--error');
    });
  });

  describe('optionalText', () => {
    it('should have element and correct classname for it', () => {
      const { getByText } = render(
        <Textarea labelText="label" optionalText="Optional" />,
      );
      const optionalText = getByText('(Optional)');
      expect(optionalText).toHaveClass('fi-label_optional-text');
    });
  });

  describe('resize', () => {
    it('allows only vertical resizing', () => {
      const { getByRole } = render(
        <Textarea labelText="label" resize="vertical" />,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      expect(textarea).toHaveStyle('resize: vertical');
    });

    it('allows only horizontal resizing', () => {
      const { getByRole } = render(
        <Textarea labelText="label" resize="horizontal" />,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      expect(textarea).toHaveStyle('resize: horizontal');
    });

    it('allows horizontal and resizing', () => {
      const { getByRole } = render(
        <Textarea labelText="label" resize="both" />,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      expect(textarea).toHaveStyle('resize: both');
    });

    it('does not allow resizing', () => {
      const { getByRole } = render(
        <Textarea labelText="label" resize="none" />,
      );
      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      expect(textarea).toHaveStyle('resize: none');
    });
  });

  describe('name', () => {
    it('has the given name attribute', () => {
      const { getByRole } = render(
        <Textarea labelText="label" resize="vertical" name="test-name" />,
      );
      const textarea = getByRole('textbox') as HTMLInputElement;
      expect(textarea.name).toBe('test-name');
    });
  });

  describe('id', () => {
    it('has the given id', () => {
      const { getByRole } = render(
        <Textarea labelText="label" id="custom-id" />,
      );
      expect(getByRole('textbox')).toHaveAttribute('id', 'custom-id');
    });
  });

  describe('fullWidth', () => {
    it('has the required classname', () => {
      const { container } = render(<Textarea labelText="label" fullWidth />);
      expect(container.firstChild).toHaveClass('fi-textarea--full-width');
    });
  });

  describe('margin', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<Textarea labelText="" margin="xs" />);
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin prop overwritten from containerProps', () => {
      const { container } = render(
        <Textarea labelText="" margin="xs" style={{ margin: 2 }} />,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });

  describe('containerProps', () => {
    it('has the given props on the container', () => {
      const { container } = render(
        <Textarea labelText="label" style={{ width: '100px' }} />,
      );
      expect(container.firstChild).toHaveAttribute('style', 'width: 100px;');
    });
  });

  describe('Character counter', () => {
    beforeAll(() => {
      jest.useFakeTimers();
    });

    afterAll(() => {
      jest.useRealTimers();
    });

    it('should display character count', () => {
      const { container, getByRole } = render(
        <Textarea
          labelText="label"
          characterLimit={20}
          ariaCharactersRemainingText={(amount) =>
            `You have ${amount} characters remaining`
          }
          ariaCharactersExceededText={(amount) =>
            `You have ${amount} characters too many`
          }
        >
          Lorem ipsum
        </Textarea>,
      );
      expect(
        container.getElementsByClassName('fi-textarea_character-counter')
          .length,
      ).toBe(1);
      expect(
        container.getElementsByClassName('fi-textarea_character-counter')[0],
      ).toHaveTextContent('11/20');

      const textarea = getByRole('textbox') as HTMLTextAreaElement;
      fireEvent.change(textarea, {
        target: { value: 'Lorem ipsum dolor sit amet' },
      });

      expect(
        container.getElementsByClassName('fi-textarea_character-counter')[0],
      ).toHaveTextContent('26/20');
    });

    it('should have correct screen reader status text', () => {
      const { container, getByTestId } = render(
        <Textarea
          labelText="label"
          defaultValue="Lorem ipsum"
          data-testid="cc-textarea"
          characterLimit={20}
          ariaCharactersRemainingText={(amount) =>
            `You have ${amount} characters remaining`
          }
          ariaCharactersExceededText={(amount) =>
            `You have ${amount} characters too many`
          }
        />,
      );

      expect(
        container.getElementsByClassName('fi-status-text')[0].firstChild,
      ).toHaveTextContent('');

      const textInput = getByTestId('cc-textarea') as HTMLTextAreaElement;

      fireEvent.change(textInput, {
        target: { value: 'Lorem ipsum dolor sit amet' },
      });

      // Testing if the delayed update works as intended
      expect(
        container.getElementsByClassName('fi-status-text')[0].firstChild,
      ).toHaveTextContent('');

      act(() => {
        jest.advanceTimersByTime(2000);
      });

      expect(
        container.getElementsByClassName('fi-status-text')[0].firstChild,
      ).toHaveTextContent('You have 6 characters too many');
    });
  });
});
