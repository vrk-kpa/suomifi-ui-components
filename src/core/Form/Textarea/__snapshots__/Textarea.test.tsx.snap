// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot default structure should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: auto;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c5::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c5::before,
.c5::after {
  box-sizing: border-box;
}

.c3.fi-label .fi-label_label-span {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  display: inline-block;
  vertical-align: middle;
  color: hsl(0,0%,13%);
}

.c3.fi-label .fi-label_label-span .fi-label_optional-text {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c3.fi-label .fi-label_label-span .fi-tooltip {
  margin-left: 6px;
}

.c6 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,13%);
  font-size: 14px;
  line-height: 20px;
}

.c6.fi-status-text {
  display: block;
}

.c6.fi-status-text.fi-status-text--error {
  color: hsl(3,59%,43%);
}

.c6.fi-status-text .fi-icon {
  vertical-align: middle;
  -webkit-transform: translateY(-0.1em);
  -ms-transform: translateY(-0.1em);
  transform: translateY(-0.1em);
  margin-right: 5px;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(0,0%,13%);
  width: 290px;
}

.c1 .fi-label,
.c1 .fi-hint-text,
.c1 .fi-status-text {
  margin: 0;
}

.c1.fi-textarea {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1.fi-textarea .fi-textarea_bottom-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1.fi-textarea .fi-textarea_character-counter {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(0,0%,13%);
  font-size: 14px;
  line-height: 20px;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  margin-top: 5px;
}

.c1.fi-textarea .fi-textarea_character-counter.fi-textarea_character-counter--error {
  color: hsl(3,59%,43%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
}

.c1.fi-textarea .fi-textarea_statusText--has-content {
  margin-top: 5px;
}

.c1.fi-textarea .fi-textarea_textarea-element-container:focus-within {
  outline: 3px solid transparent;
  position: relative;
}

.c1.fi-textarea .fi-textarea_textarea-element-container:focus-within::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1.fi-textarea .fi-label {
  margin-bottom: 10px;
}

.c1.fi-textarea .fi-hint-text {
  margin-bottom: 10px;
}

.c1.fi-textarea .fi-textarea_textarea {
  display: block;
  resize: vertical;
  border-radius: 2px;
  border: 1px solid hsl(201,7%,58%);
  box-shadow: 0 1px 2px 0 hsla(214,100%,24%,0.1) inset;
  padding: 8px 14px 13px 10px;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  width: 100%;
  background-color: hsl(0,0%,100%);
}

.c1.fi-textarea .fi-textarea_textarea:focus {
  outline: none;
}

.c1.fi-textarea .fi-textarea_textarea::-webkit-input-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1.fi-textarea .fi-textarea_textarea::-moz-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1.fi-textarea .fi-textarea_textarea:-ms-input-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1.fi-textarea .fi-textarea_textarea::placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1.fi-textarea .fi-textarea_textarea-resize--horizontal {
  resize: horizontal;
}

.c1.fi-textarea .fi-textarea_textarea-resize--both {
  resize: both;
}

.c1.fi-textarea .fi-textarea_textarea-resize--none {
  resize: none;
}

.c1.fi-textarea .fi-status-text {
  display: block;
  line-height: 18px;
}

.c1.fi-textarea.fi-textarea--disabled {
  color: hsl(202,7%,67%);
  cursor: not-allowed;
}

.c1.fi-textarea.fi-textarea--disabled .fi-textarea_textarea {
  background-color: hsl(202,7%,97%);
}

.c1.fi-textarea.fi-textarea--error .fi-textarea_textarea {
  border: 2px solid hsl(3,59%,43%);
}

.c1.fi-textarea--full-width {
  width: 100%;
}

<div
  class="c0 fi-textarea c1"
>
  <div
    class="c2 c3 fi-label"
  >
    <label
      class="c4 fi-label_label-span"
      for="just-for-snapshot-to-be-same"
    >
      Label text
    </label>
  </div>
  <div
    class="c0 fi-textarea_textarea-element-container"
  >
    <textarea
      aria-invalid="false"
      class="c5 fi-textarea_textarea"
      id="just-for-snapshot-to-be-same"
    />
  </div>
  <div
    class="c0 fi-textarea_bottom-wrapper"
  >
    <span
      aria-atomic="true"
      aria-live="assertive"
      class="c4 c6 fi-status-text"
    />
  </div>
</div>
`;
